File: Roadmap/02 - FUNCIONES Y ALCANCE/c++/AlejandroDave.cpp
#include <iostream>
using namespace std;

void sinRet(){
    cout<<"Funcion sin parametros ni retorno"<<endl;
}

float retyPar(float a){
    return a*5.8;
}

int potencia(int a){
    return a*a;
}

int altaFun(int a, int b){
    int c = a+a;
    return c+ potencia(b);
}

int ejercicioFuncion(string a, string b){
    int s  = 0;
    for(int i=0;i<100;i++){
        if(i%3==0 && i%5 ==0){
            cout<<a<<" "<<b<<endl;
        }
        if(i%3==0){
            cout<<a<<endl;
        }
        if(i%5 ==0){
            cout<<b<<endl;
        }
        else{
            cout<<i<<endl;
            s++;
        }
        }
    return s;
    }


int main(){
     float f = retyPar(4.4);
     int af = altaFun(3,5);
     sinRet();
     cout<<"Funcion con parametro y retorno "<<f<<"\nFuncion que invoca otra funcion: "<<af<<endl;
     string a = "hola", b = "c++";
     int e = ejercicioFuncion(a,b);
     cout<<"El total de veces que aparecio el numero fue: "<<e<<endl;

return 0;}

File: Roadmap/02 - FUNCIONES Y ALCANCE/python/AlejandroDave.py
# 02 Funciones y Alcance

#Funciones sin parametros ni retorno
def noParFun():
    print("Saludos mi gente")

noParFun()

#Funciones uno o varios parametros y retorno
def potNum(num1,num2):
    return pow(num1,num2)  #Aqui utilizamos una funcion predimentada, la cual genera una potencia acorde dos numeros dados.

#Dentro de funciones podemos invocar otras funciones
def listpow(lista,pot):
    listpot = []
    for x in lista:
        listpot.append(potNum(x,pot))
    return listpot

print(listpow([1,2,3,4,5],5))

'''En Python existe un tipo de funciones nombradas "High Functions" en las cuales podemos manejar otras funciones de 'baja'
    jerarquia
'''

# En una alta funcion, podemos ingresar otra funcion como parametro, y a su vez que nos regrese una funcion como retorno.
def altaFunc(f,param):
    ret = []
    return f(param)  # Nota, al invocarlas es importante tener conocimiento sobre los parametros con los que trabaja la funcion

def funSencilla(nombre):
    return "Hola {}!".format(nombre)

print(altaFunc(funSencilla,"Python"))

# Dentro de una funcion no solo podemos invocar otra funcion, si no tambien podemos crear nuevas funciones

def sumMulti(n,m):
    x = 0
    def multipl(n):  # Definimos una funcion dentro de la alta funcion
        return n*m
    for i in range(n):
        x = x + multipl(i) 
    return x

print(sumMulti(5,7))

def funcReto(par1,par2):
    for i in range(100):
        if i%3 == 0 and i%5 == 0:
            print("{} {}".format(par1, par2))
        elif i%3 == 0:
            print(par1)
        elif i%5 == 0:
            print(par2)
        else: print(i)

print(funcReto("Hola","Python"))